### PR TITLE
Moved from Scanner to bufio#ReadLine to handle larger result sets

### PR DESCRIPTION
**Pull request by smartrevolution**
Hi guys,

is this okay for you?! Tests are running, too. :)

'(stefan)

### CODE REVIEW

File: watch.go [hunk 1]
```diff
@@ -112,6 +143,9 @@ func (fb *Firebase) Watch(notifications chan Event) error {
 				// This event will be sent when the supplied auth parameter is no longer valid
 
 				// TODO: handle
+			case "rules_debug":
```
Comment (zabawaba99) [inline watch.go:152]: When do we get `rules_debug` down through sse?

File: watch.go [hunk 2]
```diff
@@ -70,11 +69,43 @@ func (fb *Firebase) Watch(notifications chan Event) error {
...
+				result = append(result, dat...)
+				if !isPrefix {
+					break
+				}
+			}
+			// Again we add the \n
+			result = append(result, '\n')
+			_, _, scanResult = scanner.ReadLine()
+			if scanResult != nil {
+				log.Println(scanResult)
```
Comment (zabawaba99) [inline watch.go]: Might be a good idea to `break` here. If we fall in here, I would assume this indicates that the connection has been killed.

File: watch.go [hunk 3]
```diff
@@ -70,11 +69,43 @@ func (fb *Firebase) Watch(notifications chan Event) error {
...
+			// 1. step: scan for the 'event:' part. ReadLine() oes not return the \n
+			// so we have to add it to our result buffer.
+			evt, isPrefix, scanResult = scanner.ReadLine()
+			result = append(result, evt...)
+			result = append(result, '\n')
+
+			// 2. step: scan for the 'data:' part. Firebase returns just one 'data:'
+			// part, but the value can be very large. If we exceed a certain length
+			// isPrefix will be true until all data is read.
+			for {
```
Comment (zabawaba99) [inline watch.go:96]: Might want to do

``` go
for !isPrefix {
   dat, isPrefix, scanResult = scanner.ReadLine()
   result = append(result, dat...)
}
```

so that we don't call `scanner.ReadLine()` another time if `isPrefix` is already false from L#86

File: watch.go (hunk 1 above)
Comment (smartrevolution) [inline watch.go:152]: When you create your own auth-token (i.e. with fireauth) with a "Debug: true" property. Sorry, I added this to this commit accidentally. I should have put this in another commit and log the data part. I suggest I add the logging to my next commit.

File: watch.go (hunk 2 above)
Comment (smartrevolution) [inline watch.go]: Yes, I could add an error handler after all ReadLine calls and break if scanResult != nil. It works this way, too. But breaking early is a good idea.

File: watch.go (hunk 3 above)
Comment (smartrevolution) [inline watch.go:96]: In L#86 we should never have isPrefix == true, because the event-type part should never be that large. Only the data part can be very large and that is why we have to call ReadLine in a loop until isPrefix == false. I did not like the idea to use isPrefix as a condition in the for-loop, as I would have to set it explicitly to isPrefix = true before the for-loop, because it is guaranteed to be false after the first and only call to ReadLine to read the event-type part. So we have to run the loop at least one time anyway. And it will stop at the right moment (when isPrefix == false). So I would leave it like this.

**Comment by smartrevolution** [comment]
Hi. I added the changes based on your recommendations and my replies. I added a log statement in the "rules_debug" section and I added checks for scanResult, for the case that the connection was already closed.

File: watch.go (hunk 1 above)
Comment (zabawaba99) [inline watch.go:152]: Cool deal. Didn't know they actually send that down with a debug token. :+1:

File: watch.go (hunk 3 above)
Comment (zabawaba99) [inline watch.go:96]: Ah, you're right. It's been a while since I tinkered with this code and I forgot that SSE looked like 

```
event: 
data: {"path":"/","data":{"foo":"bar"}}
```

My bad.

**Comment by zabawaba99** [comment]
Thanks for your contribution @smartrevolution

**Comment by smartrevolution** [comment]
Thank you very much. You make me happy. ;)
